apm-agent-core/src/test/java/co/elastic/apm/agent/report/ApmServerReporterIntegrationTest.java

```diff
@@ -136,17 +164,53 @@ void testSecretToken() throws ExecutionException, InterruptedException {
             exchange.setStatusCode(200).endExchange();
         };
         reporter.report(new Transaction(tracer));
-        reporter.flush().get();
+        reporter.flush();
         assertThat(reporter.getDropped()).isEqualTo(0);
         assertThat(receivedIntakeApiCalls.get()).isEqualTo(1);
     }
 
     @Test
-    void testReportErrorCapture() throws ExecutionException, InterruptedException {
+    void testReportErrorCapture() {
         reporter.report(new ErrorCapture(tracer));
-        reporter.flush().get();
+        reporter.flush();
+        assertThat(reporter.getDropped()).isEqualTo(0);
+        assertThat(receivedIntakeApiCalls.get()).isEqualTo(1);
+    }
+
+    @Test
+    void testTimeout() {
+        doReturn(TimeDuration.of("1ms")).when(reporterConfiguration).getApiRequestTime();
+        doAnswer(invocation -> System.nanoTime()).when(clock).nanoTicks();
+        reporter.report(new Transaction(tracer));
+        await().untilAsserted(() -> assertThat(reporter.getReported()).isEqualTo(1));
         assertThat(reporter.getDropped()).isEqualTo(0);
         assertThat(receivedIntakeApiCalls.get()).isEqualTo(1);
+        assertThat(receivedEvents.get()).isEqualTo(2);
+    }
+
+    @Test
+    void testFlush() {
+        reporter.report(new Transaction(tracer));
+        assertThat(receivedEvents.get()).isEqualTo(0);
+        assertThat(reporter.flush(5, TimeUnit.SECONDS)).isTrue();
+        assertThat(reporter.getDropped()).isEqualTo(0);
+        assertThat(reporter.getReported()).isEqualTo(1);
+    }
+
+    @Test
+    void testFailingApmServer() {
+        statusCode = HttpStatus.SERVICE_UNAVAILABLE_503;
+        // try to report a few events to trigger backoff
+        reporter.report(new Transaction(tracer));
+        reporter.flush(1, TimeUnit.SECONDS);
+        reporter.report(new Transaction(tracer));
+        reporter.flush(1, TimeUnit.SECONDS);
+        reporter.report(new Transaction(tracer));
+        reporter.flush(1, TimeUnit.SECONDS);
+        reporter.report(new Transaction(tracer));
```

Comment:
Why not test with much lower timeouts? We want to see that the default case is very quick

Reply:
Yes, the default should be quick. But I want to ensure that the backoff is triggered. Waiting for longer than needed ensures this test is less flaky. If the timeout is too low, we're not waiting for the mock APM Server to respond and thus, would not trigger the backoff. I'm flushing multiple times here as the fist backoff period is 0 seconds, meaning it retries immediately.